### PR TITLE
modify save some cross shard data to db

### LINKED ISSUE #970: simplify struct CrossShardMsgInfo

### LINKED ISSUE #971: save shard data from parent shard
#### Is your feature request related to a problem? Please describe.
A clear and concise description of what the problem is. Example: I'm facing an issue when [...]

#### Describe alternatives you have considered
A clear and concise description of any alternative solutions or features you have considered.

#### Additional context
Add any other context or screenshots about the feature request.
#### Add New db
 save shard data from parent shard,like governance related data data,contract metatdata,shard chainconfig data

### LINKED ISSUE #975: simplify struct CrossShardMsgInfo
the current struct 
```golang
type CrossShardMsgInfo struct {
	FromShardID          common.ShardID
	MsgHeight            uint32
	SignMsgHeight        uint32
	PreCrossShardMsgHash common.Uint256
	CrossShardMsgRoot    common.Uint256
	ShardMsgHashs        []*CrossShardMsgHash
}

type CrossShardMsgHash struct {
	ShardID common.ShardID
	MsgHash common.Uint256
	SigData map[uint32][]byte
}
```
should be simplified to :
```golang
type CrossShardMsgInfo struct {
	SignMsgHeight        uint32
	PreCrossShardMsgHash common.Uint256
	Index                uint32 // current shard msg hash insert pos in ShardMsgHashes
	ShardMsgHashes       []common.Uint256
	SigData              map[uint32][]byte
}
```

### LINKED ISSUE #977: ContractEvent rename to DeployContractEvent
type ContractEvent struct {
	Version  uint32
	Height   uint32
	Contract *payload.DeployCode
} 
rename to 
type DeployContractEvent struct {
	Version  uint32
	Height   uint32
	Contract *payload.DeployCode
}

### PR DESCRIPTION
**Pull request by xiemylogos**
1、save contract metadata,contract meta code to event db
2、save  chainconfig data to event db
3、refactoring  cross shard msg data structure
close #977
close #975
close #971
close #970

### CODE REVIEW

File: consensus/solo/solo.go [hunk 1]
```diff
@@ -198,67 +197,80 @@ func (self *SoloService) broadCrossShardHashMsgs(blkNum uint32, shardMsgs []xsha
...
-		crossShardMsgs.CrossMsgs = append(crossShardMsgs.CrossMsgs, crossShardMsg)
-		crossShardMsgs.Height = blkNum
-	}
-	for _, crossShard := range crossShardMsgs.CrossMsgs {
-		hashes = append(hashes, crossShard.MsgHash)
+	for _, shardMsgs := range shardMsgMap {
+		msgHash := xshard_types.GetShardCommonMsgsHash(shardMsgs)
+		hashes = append(hashes, msgHash)
 	}
 	msgRoot := common.ComputeMerkleRoot(hashes)
```
Comment (laizy) [inline consensus/solo/solo.go]: need sort the hashes by shardid, or else the msgRoot is nondeterministic

File: consensus/solo/solo.go [hunk 2]
```diff
@@ -198,67 +197,80 @@ func (self *SoloService) broadCrossShardHashMsgs(blkNum uint32, shardMsgs []xsha
...
-	for _, crossMsg := range crossShardMsgs.CrossMsgs {
-		shardMsg, present := shardMsgMap[crossMsg.ShardID]
+	sig, err := signature.Sign(self.Account, msgRoot[:])
+	if err != nil {
+		log.Errorf("sign cross shard msg root failed,msg hash:%s,err:%s", msgRoot.ToHexString(), err)
+		return
+	}
+	shardHashMap := make(map[common.Uint256]common.ShardID)
+	for shardID, shardMsgs := range shardMsgMap {
+		msgHash := xshard_types.GetShardCommonMsgsHash(shardMsgs)
```
Comment (laizy) [inline consensus/solo/solo.go]: this for loop can be merged with the upper one.

File: consensus/solo/solo.go [hunk 3]
```diff
@@ -198,67 +197,80 @@ func (self *SoloService) broadCrossShardHashMsgs(blkNum uint32, shardMsgs []xsha
...
+	sigData[0] = sig
+	crossShardMsgHash := &types.CrossShardMsgHash{
+		ShardMsgHashs: hashes,
+		SigData:       sigData,
+	}
+	for index, msgHash := range crossShardMsgHash.ShardMsgHashs {
+		shardID, present := shardHashMap[msgHash]
 		if !present {
-			log.Errorf("broadcast cross msg not found :%v", crossMsg.ShardID)
+			log.Errorf("SendCrossShardMsgToAll shard msg hash not found :%s", msgHash.ToHexString())
```
Comment (laizy) [inline consensus/solo/solo.go]: why this case will happen?

File: consensus/solo/solo.go [hunk 4]
```diff
@@ -198,67 +197,80 @@ func (self *SoloService) broadCrossShardHashMsgs(blkNum uint32, shardMsgs []xsha
...
 			continue
 		}
+		shardMsg, present := shardMsgMap[shardID]
+		if !present {
```
Comment (laizy) [inline consensus/solo/solo.go]: ditto

**Review by laizy — CHANGES_REQUESTED** [review]

File: consensus/vbft/types.go [hunk 5]
```diff
@@ -90,13 +85,9 @@ func (blk *Block) Serialize() ([]byte, error) {
...
-	if blk.CrossMsg != nil {
+	if blk.CrossMsgHash != nil {
 		sink3 := common.NewZeroCopySink(0)
-		sink3.WriteUint32(blk.CrossMsg.Height)
-		sink3.WriteVarUint(uint64(len(blk.CrossMsg.CrossMsgs)))
-		for _, crossMsg := range blk.CrossMsg.CrossMsgs {
-			crossMsg.Serialization(sink3)
-		}
+		blk.CrossMsgHash.Serialization(sink3)
 		payload.WriteVarBytes(sink3.Bytes())
```
Comment (laizy) [inline consensus/vbft/types.go]: use `SerializeToBytes(blk.CorssMsgHash)`

File: core/store/ledgerstore/tx_handler.go [hunk 6]
```diff
@@ -121,6 +124,14 @@ func HandleDeployTransaction(store store.LedgerStore, overlay *overlaydb.Overlay
 	}
 
 	notify.Notify = append(notify.Notify, notifies...)
+	notify.Notify = append(notify.Notify,
+		&event.NotifyEventInfo{
+			ContractAddress: address,
+			States: &message.ContractEvent{
```
Comment (laizy) [inline core/store/ledgerstore/tx_handler.go]: `ContractEvent` is too general, rename  to `DeployContractEvent`.

File: core/store/ledgerstore/event_store.go [hunk 7]
```diff
@@ -208,3 +210,124 @@ func (this *EventStore) getEventNotifyByTxKey(txHash common.Uint256) []byte {
...
+
+func getContractMetaDataKey(height uint32, contractAddr common.Address) []byte {
+	key := common.NewZeroCopySink(10)
+	key.WriteByte(byte(scom.CROSS_SHARD_CONTRACT_META))
+	key.WriteUint32(height)
+	key.WriteBytes(contractAddr[:])
+	return key.Bytes()
+}
+
+func (this *EventStore) SaveContractEvent(height uint32, deployCode *payload.DeployCode) error {
```
Comment (laizy) [inline core/store/ledgerstore/event_store.go]: there is no event to save, I think rename to `SaveDeployCode` should be more reasonable.

File: consensus/vbft/types.go (hunk 5 above)
Comment (xiemylogos) [reply consensus/vbft/types.go]: ok

File: core/store/ledgerstore/tx_handler.go (hunk 6 above)
Comment (xiemylogos) [reply core/store/ledgerstore/tx_handler.go]: ok

File: core/store/ledgerstore/event_store.go (hunk 7 above)
Comment (xiemylogos) [reply core/store/ledgerstore/event_store.go]: done

**Comment by Honglei-Cong** [comment]
fix #967

File: consensus/solo/solo.go [hunk 8]
```diff
@@ -316,7 +316,7 @@ func (self *SoloService) makeBlock() (*types.Block, error) {
 
 	// get ParentHeight from chain-mgr
 	parentHeight := self.ledger.GetParentHeight()
-	if self.ledger.HasParentBlockInCache(parentHeight + 1) {
+	if self.ledger.GetParentHeight() > parentHeight {
```
Comment (laizy) [inline consensus/solo/solo.go:274]: ???

Comment (Honglei-Cong) [reply consensus/solo/solo.go:274]: if self.ledger.GetParentHeight() > lastBlock.Block.Header.ParentHeight

**Review by laizy — APPROVED** [review]

**Pull request merged**